libraries/psibase/common/include/psibase/serveGraphQL.hpp

```diff
@@ -871,134 +874,316 @@ namespace psibase
+         return b;
+      }
+
+      std::string generate_sql_query(std::optional<int32_t>     limit,
+                                     bool                       descending,
+                                     std::optional<std::string> before,
+                                     std::optional<std::string> after) const
+      {
+         std::vector<std::string> filters;
+
+         if (_condition.has_value() && !_condition->empty())
+         {
+            filters.push_back(*_condition);
+         }
+
+         if (before.has_value())
+         {
+            filters.push_back(std::format("ROWID < {}", extract_cursor(before)));
+         }
+
+         if (after.has_value())
+         {
+            filters.push_back(std::format("ROWID > {}", extract_cursor(after)));
+         }
+
+         auto order = descending ? "DESC" : "ASC";
+         auto query = std::format("SELECT ROWID, * FROM \"{}\"", _table_name);
+
+         if (!filters.empty())
+         {
+            query += " WHERE ";
+            bool first = true;
+            for (const auto& filter : filters)
+            {
+               if (!first)
+                  query += " AND ";
+               query += filter;
+               first = false;
+            }
+         }
+
+         query += std::format(" ORDER BY ROWID {}", order);
+
+         if (limit.has_value())
+         {
+            query += std::format(" LIMIT {}", limit.value());
+         }
+
+         return query;
+      }
+
+      static std::string sql_query(const std::string& query, bool debug)
+      {
+         if (debug)
+         {
+            printf("[EventQuery] SQL query str: %s\n", query.c_str());
+         }
+
+         auto json_str = to<EventQueryInterface>("r-events"_a).sqlQuery(query);
+
+         if (debug)
+         {
+            printf("[EventQuery] Raw JSON response: %s\n", json_str.c_str());
+         }
+
+         return json_str;
+      }
+
+      std::string                _table_name;
+      std::optional<std::string> _condition;
+      std::optional<int32_t>     _first;
+      std::optional<int32_t>     _last;
+      std::optional<std::string> _before;
+      std::optional<std::string> _after;
+      bool                       _debug = false;
 
-      friend auto get_type_name(EventQuery*) { return psio::get_type_name<Events>(); }
    };  // EventQuery
 
+   template <typename T>
+   void from_json(SqlRow<T>& s, auto& stream)
+   {
+      from_json_object(stream,
+                       [&](std::string_view key)
+                       {
+                          if (key == "rowid")
+                          {
+                             std::string rowid_str;
+                             from_json(rowid_str, stream);
+                             auto [ptr, ec] = std::from_chars(
+                                 rowid_str.data(), rowid_str.data() + rowid_str.size(), s.rowid);
+                             check(ec == std::errc{}, "Invalid rowid");
```

Comment:
Does `from_json(s.rowid, stream)` not work?

Reply:
Doesn't this way make more sense since the json has the `rowid` value quoted? Or would you expect `from_json` to still be able to extract that into a `uint64_t`?

Reply:
Doesn't this way make more sense because the value of `rowid` is quoted? Or would you expect `from_json` to still be able to extract that into a `uint64_t`?

Reply:
`from_json` for integer types can always parse a string.